module/zfs/dmu.c

```diff
@@ -1042,245 +1277,599 @@ dmu_read_by_dnode(dnode_t *dn, uint64_t offset, uint64_t size, void *buf,
+	 * (dnode_block_freed()).  This behavior is verified by
+	 * zio_done(), which VERIFYs that the override BP is identical
+	 * to the on-disk BP.
+	 */
+	DB_DNODE_ENTER(db);
+	dn = DB_DNODE(db);
+	if (dr_next != NULL || dnode_block_freed(dn, db->db_blkid))
+		zp.zp_nopwrite = B_FALSE;
+	DB_DNODE_EXIT(db);
+
+	ASSERT(dr->dr_txg == txg);
+	if (dr->dt.dl.dr_override_state == DR_IN_DMU_SYNC ||
+	    dr->dt.dl.dr_override_state == DR_OVERRIDDEN) {
+		/*
+		 * We have already issued a sync write for this buffer,
+		 * or this buffer has already been synced.  It could not
+		 * have been dirtied since, or we would have cleared the state.
+		 */
+		mutex_exit(&db->db_mtx);
+		return (SET_ERROR(EALREADY));
+	}
+
+	ASSERT(dr->dt.dl.dr_override_state == DR_NOT_OVERRIDDEN);
+	dr->dt.dl.dr_override_state = DR_IN_DMU_SYNC;
+	mutex_exit(&db->db_mtx);
+
+	dsa = kmem_alloc(sizeof (dmu_sync_arg_t), KM_SLEEP);
+	dsa->dsa_dr = dr;
+	dsa->dsa_done = done;
+	dsa->dsa_zgd = zgd;
+	dsa->dsa_tx = NULL;
+
+	zio_nowait(arc_write(pio, os->os_spa, txg,
+	    zgd->zgd_bp, dr->dt.dl.dr_data, DBUF_IS_L2CACHEABLE(db),
+	    &zp, dmu_sync_ready, NULL, NULL, dmu_sync_done, dsa,
+	    ZIO_PRIORITY_SYNC_WRITE, ZIO_FLAG_CANFAIL, &zb));
+
+	return (0);
 }
 
-/*
- * Read 'size' bytes into the uio buffer.
- * From object zdb->db_object.
- * Starting at offset uio->uio_loffset.
- *
- * If the caller already has a dbuf in the target object
- * (e.g. its bonus buffer), this routine is faster than dmu_read_uio(),
- * because we don't have to find the dnode_t for the object.
- */
-int
-dmu_read_uio_dbuf(dmu_buf_t *zdb, uio_t *uio, uint64_t size)
-{
-	dmu_buf_impl_t *db = (dmu_buf_impl_t *)zdb;
-	dnode_t *dn;
-	int err;
+static void
+dmu_write_direct_ready(zio_t *zio)
+{
+	dmu_sync_ready(zio, NULL, zio->io_private);
+}
+
+static void
+dmu_write_direct_done(zio_t *zio)
+{
+	dmu_sync_arg_t *dsa = zio->io_private;
+	dbuf_dirty_record_t *dr = dsa->dsa_dr;
+	dmu_buf_impl_t *db = dr->dr_dbuf;
+
+	if (abd_pages_are_stable(zio->io_abd))
+		abd_release_stable_pages(zio->io_abd);
+	abd_put(zio->io_abd);
+
+	mutex_enter(&db->db_mtx);
+	if (db->db_buf) {
+		arc_buf_t *buf = db->db_buf;
+		/*
+		 * The current contents of the dbuf are now stale.
+		 */
+		ASSERT(db->db_buf == dr->dt.dl.dr_data);
+		db->db_buf = NULL;
+		db->db.db_data = NULL;
+		dr->dt.dl.dr_data = NULL;
+		arc_buf_destroy(buf, db);
+	}
+	ASSERT(db->db.db_data == NULL);
+	db->db_state = DB_UNCACHED;
+	mutex_exit(&db->db_mtx);
+
+	dmu_sync_done(zio, NULL, zio->io_private);
+	kmem_free(zio->io_bp, sizeof (blkptr_t));
+}
+
+static boolean_t
+dmu_need_to_make_pages_stable(zio_prop_t zp)
+{
+	boolean_t make_stable = B_FALSE;
+	if (zp.zp_compress != ZIO_COMPRESS_OFF ||
+	    zp.zp_checksum != ZIO_CHECKSUM_OFF ||
+	    zp.zp_encrypt == B_TRUE)
+		make_stable = B_TRUE;
+	return (make_stable);
+}
+
+static int
+dmu_write_direct(zio_t *pio, dmu_buf_impl_t *db, abd_t *data, dmu_tx_t *tx,
+    boolean_t lustre_buf)
```

Comment:
How about making this more "generic"? lustre_buf -> stable_pages

Reply:
We should be able to remove the need for this entirely if we hold the pages in the ZPL layer as suggested in a previous comment.  Then both the POSIX and Lustre IOs will be the same in the sense that we expect the pages to always be held at this point.

Reply:
We have removed the lustre_buf stuff as the pages are now mapped in the ZPL layer. So, we always set them to stable in the event we are writing. Since we do not have to worry about this with Lustre, we no longer needed the flag at all.